services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java

```diff
@@ -404,7 +420,13 @@ private static boolean backchannelLogoutClientSession(KeycloakSession session, R
             protocol.setRealm(realm)
                     .setHttpHeaders(headers)
                     .setUriInfo(uriInfo);
-            protocol.backchannelLogout(userSession, clientSession);
+
+            /*
+             * if we want to implement the specification to 100%, we would have to return the return correctly here. But
+             * since the actual session of the client cannot be distinguished from the underlying ones we decided to
+             * leave the value as it is for the first review of the pullrequest.
+             */
+            boolean clientSessionLogout = protocol.backchannelLogout(userSession, clientSession);
```

Comment:
I think we could at least return a `504` to indicate that a client wasn't successful? Wdyt?

As per 2.8.  Back-Channel Logout Response.

Reply:
From the point of the specification we would return an an 501 (see 2.7 Back-Channel Logout Actions). The problem here is that if we only have one client session (the OPs for example) a back-channel logout wouldn't make sense. In the past this part of the code has been always triggert and if it failed no actions or returns has been performed. But now it would be nice to know if this session that gets triggert with a back-channel logout is really a session that needs to be logout out via the back-channel logout or if it's just the OP's session that will be logged out later in the logout process ( see the method "backchannelLogout(KeycloakSession session, RealmModel realm, UserSessionModel userSession, UriInfo uriInfo,
 ClientConnection connection, HttpHeaders headers, boolean logoutBroker, boolean offlineSession" Class: AuthenticationManager.java line 233). If we now trigger a back-channel logout on the ops session it would fail because no IDP is configured.

Reply:
I see. I meant `504` to indicate whether or not some of the downstream sessions were not invalidated. `501` seems to be related to a failure during when processing the logout request?

Could you elaborate a bit more about the failure if no IDP is configured, please ?

Reply:
yes, the idea from the openID specification is when an error happens in the back-channel logout happens, an 501 should be returned. I would agree that it would be better to give a more detailed feedback which session failed. 

The problem is that this "back-channel logout light" (lets call it like that) is always used when an logout has been triggert. 
Currently the protocol.backchannelLogout() gets triggert but no boolean or any response getting returned. To fulfil the openID back-channel logout specification, we would need to return an http status. Now with our change we saw in a test scenario that, when there is only one session (the OP itself) and the logout gets triggert through the @GET public logout() this line of code also gets triggert and results in a false positive test with an incorrect return. 
Because at the end the session gets removed by "session.sessions().removeUserSession(realm, userSession);" even if the back-channel logout failed. The return would be 5xx but the session had been successfully removed through the removeUserSession() method. 

Maybe the best way would be to find an attribute or something that indicates the session as a non back-channel logout session. Then it would be easy when the logout process loops over all client sessions to decide which sessions should also be logout with the back-channel process. You will find this loop in the method AuthenticationManager.backchannelLogoutAll() line 324.

But we wasn't able to find something like that. Maybe this is only an problem when the OP is an Keycloak. 
If the OP is a Keycloak it always have an back-channel logout URL and supports the back-channel logout. So the OP would try to logout it self through a back-channel logout request. Which also results in a 5xx or 4xx error.

Reply:
I think I get it, see if it makes sense:

* I guess you mean `org.keycloak.protocol.oidc.OIDCLoginProtocol#backchannelLogout`, which we could change to return some information that the logout endpoint could use to decide which status code to return. At the very end of the method call you have a call to `org.keycloak.services.managers.ResourceAdminManager#sendLogoutRequest` which could return the status code that we could then use to build a Map where the key is the clientId and the value the status code. With this information, we could decide which status code to return.

* When there is a single session (the OP itself) and the current `logout` endpoint is called, I won't care much. But for the new endpoint, the expected behavior I think would be to just remove the session without involving a new logout request.

But yeah, I do think that proper status code is important for this new endpoint (not really for the current one).

Reply:
I'm not sure about returning a JSON with clientId as Key and HTTP Status of the downstream backchannel logout as value as response body of the backchannel logout REST call.

Given the scenario you have an OP Keycloak with multiple Clients connected which are called by downstream RPs. Someone sends a backchannel logout request to the OP Keycloak. This someone does typically not know which clients are configured in the OP Keycloak. OP sends backchannel logout requests to downstream clients, let's assume one returns 200 and one 501. The respose of OP Keycloak would look something like this:

Status: `504 Gateway Timeout`

Body:
```json
{
  "client_1": 200,
  "client_2": 501
}
```

* The first problem we see is that there can be multiple sessions in each downstream client that need to be logged out with one backchannel logout request. So one status code would not be enough. We could also add session id's and so on to the json to get rid of that problem.
* The second problem is that "someone" who called the backchannel logout at the Keycloak OP does normally not know about client ids and session ids of the Keycloak OP. So that's not the kind of information that should be passed as a response right?

Our suggestion is to just return a 504 in case of a downstream client failing to logout but leaving the response body empty. What do you think?

Reply:
Yeah, let's keep 504 and no body.

Let's try to come up with something that solves the original idea as a follow-up. We can do other things here like returning a link for force/re-try logout to specific clients, etc.

Reply:
Done